pkg/provisioner/ironic/ironic.go

```diff
@@ -1319,3 +1357,102 @@ func (p *ironicProvisioner) PowerOff() (result provisioner.Result, err error) {
 
 	return result, nil
 }
+
+// A private method to set target for RAID configuration steps
+func (p *ironicProvisioner) setTargetRAIDCfg(ironicNode *nodes.Node) (err error) {
+	// Build target for RAID configuration steps
+	logicalDisks, err := p.buildTargetRAIDCfg()
+	if 0 == len(logicalDisks) || nil != err {
+		return
+	}
+	// Set target for RAID configuration steps
+	err = nodes.SetRAIDConfig(
+		p.client,
+		ironicNode.UUID,
+		nodes.RAIDConfigOpts{LogicalDisks: logicalDisks},
+	).ExtractErr()
+	if nil != err {
+		p.log.Info("failed in setting the target RAID configuration",
+			"logical disks", logicalDisks,
+			"error", err,
+		)
+	}
+	return
+}
+
+// A private method to build RAID disks
+func (p *ironicProvisioner) buildTargetRAIDCfg() (logicalDisks []nodes.LogicalDisk, err error) {
+	// If not configure raid
+	if nil == p.host.Spec.RAID || nil == p.host.Spec.RAID.Volumes || 0 == len(p.host.Spec.RAID.Volumes) {
+		return
+	}
+	// Build logicalDisks
+	var logicalDisk nodes.LogicalDisk
+	var rootVolumeFlag bool = false
+	for _, volume := range p.host.Spec.RAID.Volumes {
+		if true == volume.IsRootVolume {
+			if false == rootVolumeFlag {
+				rootVolumeFlag = true
+			} else {
+				err = errors.Errorf("most one logical disk can be set to root")
+				p.log.Info("failed in build the target RAID configuration",
+					"error", err,
+				)
+				return
+			}
+		}
+		logicalDisk = nodes.LogicalDisk{
+			RAIDLevel:             volume.Level,
+			IsRootVolume:          &volume.IsRootVolume,
+			VolumeName:            volume.Name,
+			DiskType:              volume.DiskType,
+			NumberOfPhysicalDisks: volume.NumberOfPhysicalDisks,
+		}
+		if 0 == volume.SizeGibibytes {
```

Comment:
The default for logicalDisk.SizeGB is going to be nil, so this check is a little redundant. It can be simplified to just look for the interesting case, where we do have a size provided.

```
if volume.SizeGibibytes != 0 {
    logicalDisk.SizeGB = &vollume.SizeGibibytes
}
```

Notice that the 2 fields use different naming conventions. Are they using the same units?

Reply:
I think GB as same as Gibibytes, because B == byte in usually.

Reply:
According to the [Ironic docs](https://docs.openstack.org/ironic/pike/admin/raid.html#target-raid-configuration) SizeGB is in GiB :roll_eyes:, so it's correct.